streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java

```diff
@@ -781,4 +775,47 @@ public void shouldNotAllowToAddGlobalStoreWithSourceNameEqualsProcessorName() {
             sameNameForSourceAndProcessor,
             new MockProcessorSupplier());
     }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowIfNameIsNull() {
+        new InternalTopologyBuilder.Source(null, Collections.emptySet(), null);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void shouldThrowIfTopicAndPatternAreNull() {
+        new InternalTopologyBuilder.Source("name", null, null);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void shouldThrowIfBothTopicAndPatternAreNotNull() {
+        new InternalTopologyBuilder.Source("name", Collections.emptySet(), Pattern.compile(""));
+    }
+
+    @Test
+    public void shouldCompareSourceNodeWithTopicList() {
+        final InternalTopologyBuilder.Source base = new InternalTopologyBuilder.Source("name", Collections.singleton("topic"), null);
+        final InternalTopologyBuilder.Source sameAsBase = new InternalTopologyBuilder.Source("name", Collections.singleton("topic"), null);
+        final InternalTopologyBuilder.Source differentName = new InternalTopologyBuilder.Source("name2", Collections.singleton("topic"), null);
+        final InternalTopologyBuilder.Source differentTopicList = new InternalTopologyBuilder.Source("name", Collections.emptySet(), null);
+        final InternalTopologyBuilder.Source differentTopic = new InternalTopologyBuilder.Source("name", Collections.singleton("topic2"), null);
+
+        assertThat(base, equalTo(sameAsBase));
+        assertThat(base, not(equalTo(differentName)));
+        assertThat(base, not(equalTo(differentTopicList)));
+        assertThat(base, not(equalTo(differentTopic)));
+    }
+
+    @Test
+    public void shouldCompareSourceNodeWithTopicPattern() {
+        final InternalTopologyBuilder.Source base = new InternalTopologyBuilder.Source("name", null, Pattern.compile("topic"));
+        final InternalTopologyBuilder.Source sameAsBase = new InternalTopologyBuilder.Source("name", null, Pattern.compile("topic"));
+        final InternalTopologyBuilder.Source differentName = new InternalTopologyBuilder.Source("name2", null, Pattern.compile("topic"));
```

Comment:
You basically specified the `differentName` check twice, in this and in the previous method. I would specify the following tests into separate methods:
- same
- different names
- different topics
- different patterns
Makes the test better readable and avoids repetitions of checks.